src/model/names.ml

```diff
@@ -0,0 +1,70 @@
+module type Name = sig
+
+    type t
+
+    val to_string : t -> string
+
+    val of_string : string -> t
+
+    val of_ident : Ident.t -> t
+
+    val equal : t -> t -> bool
+
+    val is_hidden : t -> bool
+
+end
+
+module Name : Name = struct
+
+    type t = string
```

Comment:
No, the representation most likely won't change. The value here is that they're all abstract, so you can never confuse a `ModuleName` with a `ModuleTypeName`, which we'd almost never want to do. The interesting exception being for `Reference`s where we have more relaxed rules, so for example, that when talking about a `Constructor` we might actually want to refer to an `Exception`, but there we can consciously decided to do the conversion:

```ocaml
| `Extension -> `Resolved (`Extension(r, (
       ExtensionName.of_string (ConstructorName.to_string name)
     )))
```
It'd be a bit nicer actually to have the very few of these that we need in `names.ml`, e.g.:
```ocaml
ExtensionName.of_constructor_name : ConstructorName.t -> ExtensionName.t
```

Reply:
We had multiple bugs in the past that would have been prevented by this change, so I'm in favour -- unsurprising since I was the one who suggested it. More broadly its a bit ridiculous to go to the trouble of having accurate representations of the various paths and then just use strings for their contents.